stackrox/stackrox
# ROX-12106: Replace FormMultiSelect with ToggleGroup for Type field in image integration forms

## Description

Prerequisite to support robot accounts as credentials for Quay image integrations.

### Goal

If a new integration starts with a valid type (for example, **Registry**) instead of the current invalid empty state, then it takes only one change to select a different desired type (for example, **Scanner** or **Registry + Scanner**). In that case, there is only one change in the credentials area of the form.

### Overview

```ts
export type ImageIntegrationCategory = 'REGISTRY' | 'SCANNER' | 'NODE_SCANNER';
```

* Clairify: `categories: ('NODE_SCANNER' | 'SCANNER')[]`
* Google: `categories: ('REGISTRY' | 'SCANNER')[]`
* Quay: `categories: ('REGISTRY' | 'SCANNER')[]`

Altough toggle group which has 3 mutually exclusive items is more intuitive for user, it requires more careful implementation than multi-select list.

Match `categories` array in form state compared to:
* array which corresponds to each item according to **deep equality** comparison
* array which corresponds to item with both categories in **either order** like `['REGISTRY', 'SCANNER']` or `['SCANNER', 'REGISTRY']`

Dave, critique is especially welcome whether or not the TypeScript is fluent.

### Changed files

1. Edit cypress/helpers/formHelpers.js
    * Export `getToggleGroupItem` function

2. Edit cypress/integration/integrations/imageIntegrations.test.js
    * Replace `getSelectButtonByLabel` and `getSelectOption` with `getToggleGroupItem` calls.
    * Change assumption about default from empty categories to initially selected category.

3. Move src/Components/FormMultiSelect.tsx to PatternFly folder to be more visible for possible future use.

4. Edit 3 files in src/Containers/Integrations/IntegrationForm/Forms/QuayIntegrationForm.tsx folder:
    * Use `validCategories` in validation schema.
    * Provide a valid `categories` value in `defaultValues` object
    * Replace `FormMultiSelect` element with `ToggleGroup` element

5. Edit src/Containers/Integrations/utils/integrationUtils.ts
    * Export `categoriesUtilsForClairifyScanner` and `categoriesUtilsForRegistryScanner` objects.

6. Edit style files so disabled toggle group item matches disabled input element.
    * acs.css: Add style rule so background color becomes consistent and text has more readable contrast.
    * trumps.css: Override classic half opacity.

### Residue

1. Verify intuitive initially selected default category, or categories.
2. Import `getCategoriesText` utility function for **Type** column in integrations tables:
    * StackRox Scanner
    * Google Container Registry
    * Red Hat Quay.io
3. Rewrite `getToggleGroupItem` helper function to use `for` attribute of label and `id` attribute of toggle group?

## Checklist

## Testing Performed

### manual testing

1. `yarn lint` in ui
2. `yarn build` in ui
3. `yarn start` in ui

    * Visit **StackRox Scanner** table

    * Click link to see read-only page for existing integration
        ![clairify-view](https://user-images.githubusercontent.com/11862657/183501915-32656533-e1db-4bab-9feb-8b83851526ad.png)

    * Click **Edit** to see form, and then click **Cancel**
        ![clairify-form](https://user-images.githubusercontent.com/11862657/183501945-38f9c62b-b2fa-4d66-8403-b22d08d68f4d.png)

    * Create **New integration** to see form with default values, and then click **Cancel**
        ![clairify-new](https://user-images.githubusercontent.com/11862657/183501975-a3e6f613-6e2e-4c06-a9de-fc026db5dd5c.png)

    * Visit **Red Hat Quay.io** table

    * Click link to see read-only page for existing integration
        ![quay-view](https://user-images.githubusercontent.com/11862657/183502053-4c8e7172-f992-4ae1-9705-3997fe2fe94e.png)

    * Click **Edit** to see form, and then click **Cancel**
        ![quay-form](https://user-images.githubusercontent.com/11862657/183502086-8031f8a9-bb8e-4dca-915a-02b9bac130a8.png)

    * Create **New integration** to see form with default values, click a different value, and then click **Cancel**
        ![quay-new](https://user-images.githubusercontent.com/11862657/183502112-fba40038-c49f-45f9-866e-b642e76a0fc8.png)

    * Visit **Google Container Registry** table

    * Click **New integration** to see form with default values, and then click **Cancel**
        ![google-new](https://user-images.githubusercontent.com/11862657/183502144-901c1234-81f5-4765-9f88-181a4138aff8.png)

### integration testing

1. `yarn cypress-open` in ui/apps/platform
    * integrations/imageIntegrations.test.js
